engine/consensus/dkg/messaging_engine.go

```diff
@@ -2,185 +2,244 @@ package dkg
+	ready()
 
-func (e *MessagingEngine) forwardOutgoingMessages() {
+	done := ctx.Done()
 	for {
 		select {
-		case msg := <-e.tunnel.MsgChOut:
-			e.forwardOutboundMessageAsync(msg)
-		case <-e.unit.Quit():
+		case <-done:
 			return
+		case message := <-e.tunnel.MsgChOut:
+			go e.forwardOutboundMessage(ctx, message)
 		}
 	}
 }
 
-// forwardOutboundMessageAsync asynchronously attempts to forward a private
-// DKG message to a single other DKG participant, on a best effort basis.
-func (e *MessagingEngine) forwardOutboundMessageAsync(message msg.PrivDKGMessageOut) {
-	e.unit.Launch(func() {
-		backoff := retry.NewExponential(retryBaseWait)
-		backoff = retry.WithMaxRetries(retryMax, backoff)
-		backoff = retry.WithJitterPercent(retryJitterPct, backoff)
-
-		attempts := 1
-		err := retry.Do(e.unit.Ctx(), backoff, func(ctx context.Context) error {
-			err := e.conduit.Unicast(&message.DKGMessage, message.DestID)
-			if err != nil {
-				e.log.Warn().Err(err).Msgf("error sending dkg message retrying (%d)", attempts)
-			}
-
-			attempts++
-			return retry.RetryableError(err)
-		})
-
-		// Various network conditions can result in errors while forwarding outbound messages.
-		// Because the overall DKG is resilient to individual message failures most of time.
-		// it is acceptable to log the error and move on.
+// forwardOutboundMessage transmits message to the target DKG participant.
+// Upon any error from the Unicast, we will retry with an exponential backoff.
+// After a limited number of attempts, we will log an error and exit.
+// The DKG protocol tolerates a number of failed private messages - these will
+// be resolved by broadcasting complaints in later phases.
+// ust be invoked as a goroutine.
+func (e *MessagingEngine) forwardOutboundMessage(ctx context.Context, message msg.PrivDKGMessageOut) {
+	backoff := retry.NewExponential(e.config.RetryBaseWait)
+	backoff = retry.WithMaxRetries(e.config.RetryMax, backoff)
+	backoff = retry.WithJitterPercent(e.config.RetryJitterPercent, backoff)
+
+	started := time.Now()
+	log := e.log.With().Str("target", message.DestID.String()).Logger()
+
+	attempts := 0
+	err := retry.Do(ctx, backoff, func(ctx context.Context) error {
+		attempts++
+		err := e.conduit.Unicast(&message.DKGMessage, message.DestID)
+		// TODO Unicast does not document expected errors, therefore we treat all errors as benign networking failures here
 		if err != nil {
-			e.log.Error().Err(err).Msgf("error sending private dkg message after %d attempts", attempts)
+			log.Warn().
+				Err(err).
+				Int("attempt", attempts).
+				Dur("send_time", time.Since(started)).
+				Msgf("error sending dkg message on attempt %d - will retry...", attempts)
 		}
+
+		return retry.RetryableError(err)
 	})
+
+	// TODO Unicast does not document expected errors, therefore we treat all errors as benign networking failures here
+	if err != nil {
```

Comment:
I would assume `retry.Do` blocks before it sends the message, doesn't it?

Reply:
On the first attempt, it will send the message immediately. On subsequent attempts, it will sleep based on the backoff configuration.